base/strings/search.jl

```diff
@@ -166,17 +188,21 @@ function _search_bloom_mask(c)
 end
 
 _nthbyte(s::String, i) = codeunit(s, i)
-_nthbyte(a::Union{AbstractVector{UInt8},AbstractVector{Int8}}, i) = a[i]
+_nthbyte(t::AbstractVector, index) = t[index + (firstindex(t)-1)]
 
 function _searchindex(s::String, t::String, i::Integer)
     # Check for fast case of a single byte
     lastindex(t) == 1 && return something(findnext(isequal(t[1]), s, i), 0)
     _searchindex(unsafe_wrap(Vector{UInt8},s), unsafe_wrap(Vector{UInt8},t), i)
 end
 
-function _searchindex(s::ByteArray, t::ByteArray, i::Integer)
-    n = sizeof(t)
-    m = sizeof(s)
+function _searchindex(s::AbstractVector{<:Union{Int8,UInt8}},
+                      t::AbstractVector{<:Union{Int8,UInt8}},
+                      _i::Integer)
+    n = length(t)
+    m = length(s)
+    i = Int(_i) - (firstindex(s) - 1)
+    i < 1 && throw(BoundsError(s, _i))
```

Comment:
Should this be `(i < 1 || i > m+1)`?

Reply:
See also below.

Presumably there are also bounds checks in the `String` case; it would be nice to combine the two bounds checks into `_searchindex`  if possible.

Reply:
the difficulty is that when `String` is not `ASCII`, the bounds is difficult to check but I will give it a try

Reply:
I don't think it matters — this is used for `String`, where the indices are code units (i.e. bytes), and the current behavior is for `findnext` is to throw `BoundsError` for indices `> m+1` (i.e. the indices can be one byte past the end of the UTF-8 data but not farther).

e.g. `findnext("a", "α", 3)` is allowed but `findnext("a", "α", 4)` throws an error.

So `(i < 1 || i > m+1)` should be correct for `String` searches as well.

Reply:
ah, not quite, because eventually the string version boils down to:
https://github.com/JuliaLang/julia/blob/41e603e5836bebf447f14c82fb2671b3567fc3bf/base/strings/search.jl#L85

I can add a check there maybe?

Reply:
```
julia> findprev("dha", "askdha", 100)
4:6
```
found out this is actually valid......time to re-write test

Reply:
@goretkin Does this sound right to you given what we discussed at #36967?